0.preprocess-sites/8.visualize-plate-layout.py

```diff
@@ -0,0 +1,392 @@
+by_well_gg.save(output_file, dpi=300, verbose=False)
+
+# Plot cell category ratios per well
+ratio_df = pd.pivot_table(
+    cell_count_df,
+    values="cell_count",
+    index=["site", "Plate", "Well", "Site", "x_loc", "y_loc"],
+    columns=["Cell_Quality"],
+)
+ratio_df["Sum"] = ratio_df.sum(axis=1)
+ratio_df["Pass_Filter"] = ratio_df[cell_filter].sum(axis=1)
+fail_filter = [cat for cat in cell_category_order if cat not in cell_filter]
+ratio_df["Fail_Filter"] = ratio_df[fail_filter].sum(axis=1)
+fail_filter_noempty = fail_filter.remove("Empty")
+ratio_df["Fail_Filter_noempty"] = ratio_df[fail_filter].sum(axis=1)
+ratio_df["Pass/Fail_withempty"] = ratio_df["Pass_Filter"] / ratio_df["Fail_Filter"]
+ratio_df["Pass/Fail_0empty"] = ratio_df["Pass_Filter"] / ratio_df["Fail_Filter_noempty"]
+ratio_df["Percent_Empty"] = ratio_df["Empty"] / ratio_df["Sum"]
+ratio_df = ratio_df.drop(
+    cell_category_order + ["Sum", "Pass_Filter", "Fail_Filter", "Fail_Filter_noempty"],
+    1,
+)
+ratio_df = ratio_df.stack().to_frame().reset_index().rename(columns={0: "value"})
+
+ratio_gg = (
+    gg.ggplot(ratio_df, gg.aes(x="x_loc", y="y_loc"))
+    + gg.geom_point(gg.aes(fill="value"), size=10)
+    + gg.geom_text(gg.aes(label="Site"), color="lightgrey")
+    + gg.facet_grid("Cell_Quality~Well", scales="free_y")
+    + gg.coord_fixed()
+    + gg.theme_bw()
+    + gg.theme(
+        axis_text=gg.element_blank(),
+        axis_title=gg.element_blank(),
+        strip_background=gg.element_rect(colour="black", fill="#fdfff4"),
+    )
+    + gg.scale_fill_cmap(name="magma")
+)
+
+output_file = pathlib.Path(output_figuresdir, "plate_layout_ratios_per_well.png")
+ratio_gg.save(output_file, dpi=300, verbose=False)
+
+# Create image dataframe
+sites = [x for x in os.listdir(batch_dir) if x not in ignore_files]
+image_list = []
+
+for site in sites:
+    try:
+        print(f"Now processing {site}...")
+        image_file = pathlib.Path(batch_dir, site, "Image.csv")
+        # Aggregates image information by site into single list
+        image_df = pd.read_csv(image_file)
+        image_df["site"] = site
+        image_df = image_df.assign(
+            Plate=[x[0] for x in image_df.site.str.split("-")],
+            Well=[x[1] for x in image_df.site.str.split("-")],
+            Site=[x[2] for x in image_df.site.str.split("-")],
+        )
+        image_list.append(image_df)
+    except FileNotFoundError:
+        print(f"{site} data not found")
+        continue
+
+image_df = pd.concat(image_list, axis="rows").reset_index(drop=True)
+print("Done concatenating image files")
+
+# add in x, y coordinates for plotting
+image_df["Site"] = image_df["Site"].astype(int)
+loc_df["Site"] = loc_df["Site"].astype(int)
+image_df = image_df.merge(loc_df, how="left", on="Site")
+
+# Plot final Cells thresholds per well
+cells_finalthresh_gg = (
+    gg.ggplot(image_df, gg.aes(x="x_loc", y="y_loc"))
+    + gg.geom_point(gg.aes(fill="Threshold_FinalThreshold_Cells"), size=10)
```

Comment:
I am a bit concerned here (and in many subsequent lines) of the hardcoding of `Threshold_FinalThreshold_Cells`. Are you confident that this column name will never change?

We might want to consider creating a separate QC yaml file that we can include all of these column names and then (instead of adding more figures here, loop over the desired QC columns)

Reply:
For plotting the compartment thresholds it would be good to have this rewritten such that it iterates through all compartments (that are not Cytoplasm) instead of just Nuclei and Cells. (Right now Nuclei and Cells are our only non-Cytoplasm compartments but when we eventually use SABER there will be more).
I don't think it needs to be done right now as SABER is a while off.

Threshold_FinalThreshold_{compartment} is unlikely to ever change, but could in theory during a major CellProfiler release (I believe it would be considered a breaking change) so we could certainly pull out the column name.

Reply:
(I guess that is assuming that when we do SABER we actually identify objects in the additional stains/compartments. I don't actually know that is true.)